inc/lift/SharedRequest.h

```diff
@@ -0,0 +1,41 @@
+#pragma once
+
+#include <memory>
+
+namespace lift {
+
+class RequestPool;
+class Request;
+
+class SharedRequest
+{
+public:
+    SharedRequest(RequestPool& request_pool, std::unique_ptr<Request> request_handle);
+    SharedRequest(const SharedRequest&) = delete;
+    SharedRequest(SharedRequest&&) = delete;
+    auto operator=(const SharedRequest&) = delete;
+    auto operator=(SharedRequest&&) = delete;
+    ~SharedRequest();
+    
+    /**
+     * @return Access to the underlying request handle.
+     * @{
+     */
+    [[nodiscard]]
+    auto GetAsReference() -> Request& { return *m_request; }
+    [[nodiscard]]
+    auto GetAsReference() const -> const Request& { return *m_request; }
+    [[nodiscard]]
+    auto GetAsPointer() -> Request* { return m_request.get(); }
+    [[nodiscard]]
+    auto GetAsPointer() const -> const Request* { return m_request.get(); }
+    /** @} */
+private:
+    /// The request pool that owns this request.
+    RequestPool& m_request_pool;
+    
+    /// The actual underlying request object.
+    std::unique_ptr<Request> m_request;
+};
```

Comment:
I think you could replace SharedRequest with a unique_ptr + custom deleter (lambda capturing request_pool, which "deletes" by releasing back into the pool)

Reply:
The lambda type becomes an issue for the unique pointer template, so it could be done with a struct deleter class, but the unique pointer would then be a different type than the type in the request pool, so we'd have to release and then reacquire before moving. I don't think there's a gain by changing to a unique pointer, so I'd rather leave as is.